vowpalwabbit/core/src/reductions/cb/cb_to_cb_adf.cc

```diff
@@ -178,36 +176,66 @@ VW::LEARNER::base_learner* VW::reductions::cb_to_cb_adf_setup(VW::setup_base_i&
   }
 
   // if cb_explore_adf is being specified this is a noop
-  if (override_cbify && options.was_supplied("cb_explore_adf")) { return nullptr; }
+  if (cb_to_cb_adf_opts->override_cbify && options.was_supplied("cb_explore_adf")) { return nullptr; }
 
-  if (override_cbify)
-  {
-    options.insert("cb_explore_adf", "");
-    // no need to register custom predict/learn, cbify will take care of that
-    return stack_builder.setup_base_learner();
-  }
+  if (cb_to_cb_adf_opts->override_cbify) { options.insert("cb_explore_adf", ""); }
 
   // user specified "cb_explore" but we're not using an old model file
-  if (override_cb_explore)
+  if (cb_to_cb_adf_opts->override_cb_explore)
   {
-    num_actions = cbx_num_actions;
+    cb_to_cb_adf_opts->num_actions = cb_to_cb_adf_opts->cbx_num_actions;
     options.insert("cb_explore_adf", "");
   }
   else  // if (override_cb) case
   {
     // force cb_adf; cb_adf will pick up cb_type
     options.insert("cb_adf", "");
   }
+  return cb_to_cb_adf_opts;
+}
+
+}  // namespace
+
+/*
+    Purpose: run before cb, cb_explore, cbify and cb_adf related reductions
+    This will 'translate' cb (non adf) input commands into their cb_adf counterparts
+    Except when:
+        - the model file loaded is from a version older than and not including 8.11.0
+        - user bypasses this translation step using '--cb_force_legacy'
+
+    Related files: cb_algs.cc, cb_explore.cc, cbify.cc
+*/
+VW::LEARNER::base_learner* VW::reductions::cb_to_cb_adf_setup(VW::setup_base_i& stack_builder)
+{
+  options_i& options = *stack_builder.get_options();
+  VW::workspace& all = *stack_builder.get_all_pointer();
+  auto cb_to_cb_adf_opts = get_cb_to_cb_adf_options_instance(all, options);
+  if (cb_to_cb_adf_opts == nullptr) { return nullptr; }
+
+  if (cb_to_cb_adf_opts->cb_to_cbadf_supplied)
+  {
+    all.logger.out_warn(
+        "The flag --cb_to_cbadf has no effect and should not be supplied. The cb_to_cbadf reduction is automatically "
+        "enabled if cb, cb_explore or cbify are used. The cb_to_cbadf reduction can be force disabled with "
+        "--cb_force_legacy. This flag will be removed in a future release but not the functionality.");
+  }
+
+  if (cb_to_cb_adf_opts->override_cbify)
+  {
+    // no need to register custom predict/learn, cbify will take care of that
+    return stack_builder.setup_base_learner();
+  }
 
   auto data = VW::make_unique<cb_to_cb_adf>();
-  data->explore_mode = override_cb_explore;
+  data->explore_mode = cb_to_cb_adf_opts->override_cb_explore;
   data->weights = &(all.weights);
 
   multi_learner* base = as_multiline(stack_builder.setup_base_learner());
 
-  if (num_actions <= 0) { THROW("cb num actions must be positive"); }
+  if (cb_to_cb_adf_opts->num_actions <= 0) { THROW("cb num actions must be positive"); }
```

Comment:
should this live in get____options_instance()

Reply:
makes sense, moved